### PR TITLE
Ci mypy

### LINKED ISSUE #103: ci: add mypy

### LINKED ISSUE #104: add version to documentation requirement and combine them with requirements

### LINKED ISSUE #108: ci:test documentation before push to main

### PR DESCRIPTION
**Pull request by HLasse**
Copy-pasted from previous PR
---
fixes #104, fixes #103, fixes #108 

## Note for reviewer:
- ~~Now the extract_dict output is a jsonl style list of dicts instead of a singular dict with keys and list of values. What is the ideal format?~~
- The current tests for extract_df are quite bad (we, e.g. don't check that we get all the right keys out). I at least made plenty of mistakes during refactored which wasn't caught - might be worth adding in another PR
- Merged #110 into this branch as well as this branch fix some of the tests

### CODE REVIEW

File: .github/workflows/test_and_release.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  pull_request:
+    branches: [main, dev]
+
+jobs:
+  pytest-and-notebooks:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        python-version: ["3.9", "3.8", "3.7"]
```
Comment (HLasse) [inline .github/workflows/test_and_release.yml:14]:
```suggestion
        os: [ubuntu-latest]
        python-version: ["3.9", "3.8", "3.7"]
```
Seems a bit overkill to run on all OS's - let's remove some to save time

File: docs/quality.rst [hunk 2]
```diff
@@ -59,7 +59,58 @@ Usage
...
+
+.. code-block:: python
+
+    import spacy
+    import textdescriptives as td
+    nlp = spacy.load("en_core_web_sm")
+
+    # set thresholds for quality metrics (these are just the default)
+    thresholds = QualityThresholds(
+        n_stop_words(2, None),
```
Comment (HLasse) [inline docs/quality.rst:74]:
```suggestion
        n_stop_words=(2, None),
```

File: docs/quality.rst [hunk 3]
```diff
@@ -59,7 +59,58 @@ Usage
...
+        duplicate_8gram_chr_fraction=(None, 0.12),
+        duplicate_9gram_chr_fraction=(None, 0.11),
+        duplicate_10gram_chr_fraction=(None, 0.1),
+        top_2gram_chr_fraction=(None, 0.20),
+        top_3gram_chr_fraction=(None, 0.18),
+        top_4gram_chr_fraction=(None, 0.16),
+        contains_lorem_ipsum=False
+    )
+
+    nlp.add_pipe("textdescriptives.quality", config={"quality_thresholds": thresholds.dict()})
```
Comment (HLasse) [inline docs/quality.rst:95]:
```suggestion
    nlp.add_pipe("textdescriptives.quality", config={"quality_thresholds": thresholds.dict()})
```
Lidt ærgerligt det er nødvendigt at unpacke til en dict, men 🤷‍♂️

File: textdescriptives/components/quality.py [hunk 4]
```diff
@@ -4,50 +4,121 @@
...
+    doc_length: Interval = Field(
+        (10, 100_000),
+        description="A Range for the document length. Default: (10, 100_000), i.e."
+        + " between 10 and 100_000 characters.",
+    )
+    symbol_to_word_ratio: Dict[str, Interval] = Field(
+        {"#": (None, 0.1)},
+        description="A dict of symbols and the allowed range for the "
+        + r"symbol-to-word-ratio. Default: {'#': (None, 0.1)} i.e. no lower limit, "
+        + r"but at most 10% of words start with a hashtag. Values not in the dict "
```
Comment (HLasse) [inline textdescriptives/components/quality.py:42]: Does this check how many words start with the symbol or how many symbols there are compared to words?

File: textdescriptives/components/quality.py [hunk 5]
```diff
@@ -434,36 +509,100 @@ def quality_getter(self, span: Span) -> Dict[str, Union[float, int, bool]]:
...
+                quality[name] = getter(span)  # type: ignore
         return quality
 
+    @staticmethod
+    def is_within_range(rangetuple: Interval, value: float) -> bool:
+        """Check if a value is within a range tuple. If one of the values in
+        the range tuple is None it is considered to be unbounded.
+
+        Args:
+            rangetuple (RangeTuple): range tuple
```
Comment (HLasse) [inline textdescriptives/components/quality.py:529]:
```suggestion
            rangetuple (Interval): range tuple
```

File: textdescriptives/components/quality.py [hunk 6]
```diff
@@ -434,36 +509,100 @@ def quality_getter(self, span: Span) -> Dict[str, Union[float, int, bool]]:
...
+        if not self.is_within_range(qt.alpha_ratio, quality["alpha_ratio"]):
+            return False
+        if not self.is_within_range(qt.mean_word_length, quality["mean_word_length"]):
+            return False
+        if not self.is_within_range(qt.doc_length, quality["doc_length"]):
+            return False
+        if not self.is_within_range(
+            qt.proportion_ellipsis,
+            quality["proportion_ellipsis"],
+        ):
```
Comment (HLasse) [inline textdescriptives/components/quality.py:556]: I feel like we could loop this.. zip the qt.X and keys?

**Review by HLasse — COMMENTED** [review]
Great work, @KennethEnevoldsen ! Only some minor points otherwise feel free to merge in

File: .github/workflows/test_and_release.yml (hunk 1 above)
Comment (KennethEnevoldsen) [reply .github/workflows/test_and_release.yml:14]: It runs in parallel? We could get issues with windows; I wouldn't want to remove that (then I would rather remove 3.8), but I would rather just run it on all. If speed is an issue, we could ensure that the test does not need to download a spacy pipeline.

File: textdescriptives/components/quality.py (hunk 4 above)
Comment (KennethEnevoldsen) [reply textdescriptives/components/quality.py:42]: It calculates the ratio between symbol occurrence and word occurrence and checks whether it is below a threshold  0.1. (so if we have 100 words the symbol should appear no more than 10 times).

File: textdescriptives/components/quality.py (hunk 6 above)
Comment (KennethEnevoldsen) [reply textdescriptives/components/quality.py:556]: We could. I simply liked that it was explicit about how each key is treated.

**Comment by KennethEnevoldsen** [comment]
Test fails due as tutorials needs to be updated.

File: .github/workflows/test_and_release.yml (hunk 1 above)
Comment (HLasse) [reply .github/workflows/test_and_release.yml:14]: Fair - let's go with the whole grid

File: textdescriptives/components/quality.py (hunk 4 above)
Comment (HLasse) [reply textdescriptives/components/quality.py:42]: can you update the description to reflect that? (copy paste the above is fine)

File: textdescriptives/components/quality.py (hunk 6 above)
Comment (HLasse) [reply textdescriptives/components/quality.py:556]: No strong opinion on my end, let's keep it as is

**Pull request merged**